Incorrect IDE0047 Parentheses can be removed diagnostic

**Version Used**: 4e4c823

[SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKBQZgAJ1CBhQgb10JuKJQBZCBZACgHtgArOAYxkIAPAJSVqtCQEsAZoVaDCkqHID6w4eIk0qOLbQC+mwoZz6gA===)

```cs
using System;
public class C {
    public void M(object x) {
        if (x is (_)) // info IDE0047: Parentheses can be removed
        {
        }
    }
}
```

**Expected Behavior**: No IDE0047 diagnostic

**Actual Behavior**: IDE0047 is given, but when the user applies the fix, it introduces a compile error `error CS0246: The type or namespace name '_' could not be found` because the expression is now considered to be an "is type" expression instead of "is pattern" and old binding rules apply.

I think this scenario may be changed if @jmarolf's suggested feature to remove non-lexical contextual keywords is implemented.

--- a/src/Analyzers/CSharp/Tests/RemoveUnnecessaryParentheses/RemoveUnnecessaryPatternParenthesesTests.cs
+++ b/src/Analyzers/CSharp/Tests/RemoveUnnecessaryParentheses/RemoveUnnecessaryPatternParenthesesTests.cs
@@ -250,5 +250,21 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.RemoveUnnecessaryParent
     }
 }", offeredWhenRequireForClarityIsEnabled: false);
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestAlwaysNecessaryForDiscard()
+        {
+            await TestDiagnosticMissingAsync(
+@"
+class C
+{
+    void M(object o)
+    {
+        if (o is $$(_))
+        {
+        }
+    }
+}");
+        }
     }
 }

--- a/src/Analyzers/CSharp/Tests/RemoveUnnecessaryParentheses/RemoveUnnecessaryPatternParenthesesTests.cs
+++ b/src/Analyzers/CSharp/Tests/RemoveUnnecessaryParentheses/RemoveUnnecessaryPatternParenthesesTests.cs
@@ -12,6 +12,7 @@ using Microsoft.CodeAnalysis.CSharp.RemoveUnnecessaryParentheses;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Diagnostics;
 using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
 using Xunit;
 using Xunit.Abstractions;
 
@@ -251,6 +252,7 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.RemoveUnnecessaryParent
 }", offeredWhenRequireForClarityIsEnabled: false);
         }
 
+        [WorkItem(52589, "https://github.com/dotnet/roslyn/issues/52589")]
         [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
         public async Task TestAlwaysNecessaryForDiscard()
         {
@@ -265,6 +267,32 @@ class C
         }
     }
 }");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestUnnecessaryForDiscardInSubpattern()
+        {
+            await TestAsync(
+@"
+class C
+{
+    void M(object o)
+    {
+        if (o is string { Length: $$(_) })
+        {
+        }
+    }
+}",
+@"
+class C
+{
+    void M(object o)
+    {
+        if (o is string { Length: _ })
+        {
+        }
+    }
+}", offeredWhenRequireForClarityIsEnabled: true);
         }
     }
 }

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
@@ -689,6 +689,11 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
             if (pattern is ParenthesizedPatternSyntax)
                 return true;
 
+            // We're parenthesized discard pattern. We cannot remove parens.
+            // x is (_)
+            if (pattern is DiscardPatternSyntax {Parent: ParenthesizedPatternSyntax {Parent: IsPatternExpressionSyntax}})
+                return false;
+
             // (not ...) -> not ...
             //
             // this is safe because unary patterns have the highest precedence, so even if you had:

--- a/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
+++ b/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/ParenthesizedExpressionSyntaxExtensions.cs
@@ -691,7 +691,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
 
             // We're parenthesized discard pattern. We cannot remove parens.
             // x is (_)
-            if (pattern is DiscardPatternSyntax {Parent: ParenthesizedPatternSyntax {Parent: IsPatternExpressionSyntax}})
+            if (pattern is DiscardPatternSyntax && node.Parent is IsPatternExpressionSyntax)
                 return false;
 
             // (not ...) -> not ...

